mlrun/runtimes/base.py

```diff
@@ -1284,6 +1284,86 @@ def monitor_runs(self, db: DBInterface, db_session: Session):
                     exc=str(exc),
                     traceback=traceback.format_exc(),
                 )
+        for project, runs in project_run_uid_map.items():
+            if runs:
+                for run_uid in runs:
+                    self._ensure_run_not_stuck_on_non_terminal_state(
+                        db, db_session, project, run_uid, runtime_resources
+                    )
+
+    def _ensure_run_not_stuck_on_non_terminal_state(
+        self, db: DBInterface, db_session: Session, project, run_uid, runtime_resources
+    ):
+        """
+        Assuring that a run does not become trapped in a non-terminal state as a result of not finding
+        corresponding k8s resource.
+        This can occur when a node is evicted or preempted, causing the resources to be removed from the resource
+        listing when the final state recorded in the database is non-terminal.
+        This will have a significant impact on scheduled jobs, since they will not be created until the
+        previous run reaches a terminal state
+        """
+        try:
+            run = db.read_run(db_session, run_uid, project)
```

Comment:
Talked offline, we might have dozens of thousands of runs, we can't hit the DB for each one, we already have `_list_runs_for_monitoring` that is listing all the runs for you, I actually don't even think you need to make sure the runs there hold the changes that possibly happened in `_monitor_runtime_resource`, the changes will simply be reflected in the next monitoring cycle